File: dataStructures/trie.rb
class Trie

  def initialize
    @count = 0
    @root = TrieNode.new
    @alphabet = ("a".."z").to_a
  end

  def insert(key)
    key = key.downcase
    pCrawl = @root
    @count += 1
    key.each_char do |c|
      ind = @alphabet.index(c)
      if (pCrawl.children[ind].nil?)
        pCrawl.children[ind] = TrieNode.new
      end
      pCrawl = pCrawl.children[ind]
    end
    pCrawl.value = @count
  end

  def search(key)
    key = key.downcase
    pCrawl = @root
    key.each_char do |c|
      ind = @alphabet.index(c)
      if (pCrawl.children[ind].nil?)
        return false
      end
      pCrawl = pCrawl.children[ind]
    end
    return pCrawl
  end

end

class TrieNode

  attr_accessor :value, :children

  def initialize
    self.value = 0
    self.children = [nil]*26
  end

end

File: dataStructures/trie.py
import string

class Trie:

    def __init__(self):
        self.count = 0
        self.root = TrieNode()

    def insert(self, key):
        key = key.lower()
        self.count += 1
        pCrawl = self.root
        for char in key:
            ind = string.lowercase.index(char)
            if (pCrawl.children[ind] is None):
                pCrawl.children[ind] = TrieNode()
            pCrawl = pCrawl.children[ind]
        pCrawl.value = self.count
    

    def search(self, key):
        key = key.lower()
        pCrawl = self.root
        for char in key:
            ind = string.lowercase.index(char)
            if (pCrawl.children[ind] is None):
                return None
            pCrawl = pCrawl.children[ind]
        return pCrawl

    def __contains__(self, key):
        if (self.search(key)):
            return True
        else:
            return False

    


class TrieNode:

    def __init__(self):
        self.value = 0
        self.children = [None]*len(string.lowercase)


t = Trie()
t.insert("karl")
t.insert("butts")
print "kar" in t
print "but" in t
